text/0000-panic-in-drop.md

```diff
@@ -0,0 +1,319 @@
+
+```rust
+impl BufWriter {
+    /// Closes the underlying file and reports any errors that occurred while
+    /// flushing the buffer.
+    ///
+    /// Note that the underlying file is still closed even if I/O errors
+    /// occurred.
+    pub fn close(self) -> std::io::Result<()> {
+        let err = self.flush();
+        drop(self);
+        err
+    }
+}
+```
+
+Note that it is still possible to catch panics inside a `Drop` impl using `catch_unwind`. It's when the panic attempts to unwind out of the `drop` function that the process is aborted.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The basic functionality behind this RFC is already implemented as an unstable compiler flag `-Z panic-in-drop={abort,unwind}`. This RFC proposes to change the default to `abort` instead of `unwind`.
+
+This flag modifies the behavior of the `drop_in_place<T>` intrinsic function, which is responsible for calling the `drop` function of a type (if it has one) and then recursively calling `drop_in_place` for all its fields. Specifically:
+- rustc recognizes `drop_in_place` as a function that can never unwind and therefore avoids generating unnecessary landing pads when calling it.
+- Within `drop_in_place`, rustc inserts an abort guard to catch any unwinds escaping a call to `Drop::drop` to turn them into aborts.
+- The `drop_in_place` function is marked with the `nounwind` LLVM attribute, which allows LLVM to optimize away unnecessary landing pads.
+
+This is best shown in this pseudo-rust code:
+
+```rust
+// Codegen with -Z panic-in-drop=unwind (old behavior)
+unsafe fn drop_in_place<T>(ptr: *mut T) {
+    // Call the Drop impl if there is one.
+    try {
+        <T as Drop>::drop(&mut *ptr);
+    } catch {
+        // If the Drop impl panics the keep trying to drop fields.
+        try {
+            drop_in_place(&mut (*ptr).field1);
+            drop_in_place(&mut (*ptr).field2);
+            drop_in_place(&mut (*ptr).field3);
+        } catch {
+            abort();
+        }
+    }
+
+    // Drop the first field.
+    try {
+        drop_in_place(&mut (*ptr).field1);
+    } catch {
+        // If dropping field1 panics the keep trying to drop the remaining fields.
+        try {
+            drop_in_place(&mut (*ptr).field2);
+            drop_in_place(&mut (*ptr).field3);
+        } catch {
+            abort();
+        }
+    }
+
+    // Drop the second field.
+    try {
+        drop_in_place(&mut (*ptr).field2);
+    } catch {
+        // If dropping field2 panics the keep trying to drop the remaining fields.
+        try {
+            drop_in_place(&mut (*ptr).field3);
+        } catch {
+            abort();
+        }
+    }
+
+    // Drop the third field.
+    drop_in_place(&mut (*ptr).field3);
+}
```

Comment:
I realize that this is illustrative rather than accurate, but this is also *significantly* worse than what is necessary and as such exaggerates the benefit of `panic-in-drop=abort` simplifying drop glue. A better representation of code the compiler could emit would be

```rust
unsafe fn drop_in_place<T>(ptr: *mut T) {
    let mut unwind = None;

    // Call the Drop impl if there is one.
    try {
        <T as Drop>::drop(&mut *ptr);
    } catch(payload) {
        std::panicking::panic_count::increase();
        match unwind {
            Some(_) => abort(),
            None => unwind = Some(payload),
        }
    }

    // Drop the first field.
    try {
        drop_in_place(&mut (*ptr).field1);
    } catch {
        std::panicking::panic_count::increase();
        match unwind {
            Some(_) => abort(),
            None => unwind = Some(payload),
        }
    }

    // Drop the second field.
    try {
        drop_in_place(&mut (*ptr).field2);
    } catch {
        std::panicking::panic_count::increase();
        match unwind {
            Some(_) => abort(),
            None => unwind = Some(payload),
        }
    }

    // Drop the third field.
    try {
        drop_in_place(&mut (*ptr).field3);
    } catch {
        std::panicking::panic_count::increase();
        match unwind {
            Some(_) => abort(),
            None => unwind = Some(payload),
        }
    }
    
    // Resume an in-progress unwind.
    if let Some(payload) = unwind {
        std::panicking::panic_count::decrease();
        resume_unwind(payload);
    }
}
```

but even this is more complicated than it need be, as those `abort` calls could be replaced by `unreachable_unchecked`, as hitting that case requires a double panic to unwind rather than immediately abort.

<details><summary>maybe better suggestion</summary>

```suggestion
// Codegen with -Z panic-in-drop=unwind (old behavior)
unsafe fn drop_in_place<T>(ptr: *mut T) {
    let unwind;

    // Call the Drop impl if there is one.
    try {
        <T as Drop>::drop(&mut *ptr);
    } catch(payload) {
        magic::assume!(!magic::initialized!(unwind));
        std::panicking::panic_count::increase();
        unwind = payload;
    }

    // Drop the first field.
    try {
        drop_in_place(&mut (*ptr).field1);
    } catch {
        magic::assume!(!magic::initialized!(unwind));
        std::panicking::panic_count::increase();
        unwind = payload;
    }

    // Drop the second field.
    try {
        drop_in_place(&mut (*ptr).field2);
    } catch {
        magic::assume!(!magic::initialized!(unwind));
        std::panicking::panic_count::increase();
        unwind = payload;
    }

    // Drop the third field.
    try {
        drop_in_place(&mut (*ptr).field3);
    } catch {
        magic::assume!(!magic::initialized!(unwind));
        std::panicking::panic_count::increase();
        unwind = payload;
    }
    
    // Resume an in-progress unwind.
    if magic::initialized!(unwind) {
        std::panicking::panic_count::decrease();
        resume_unwind(unwind);
    }
}
```

</details>

</details>

Reply:
Actually the code that I showed is pretty much what is currently emitted by the compiler. Specifically:
- The panic count is not checked, because it requires a TLS access which can be very slow.
- Separate drop calls are used in the normal path and the unwind path because a different landing pad needs to be used when those calls unwind.

Reply:
Also, because of `extern "C-unwind"` we can have C++ exceptions unwinding through Rust code which *don't* increment the panic count. This means that we can't rely on the panic count to guarantee that a drop within an unwind guard won't itself panic.

Reply:
So the compiler is emitting `intrinsics::r#try` directly. AIUI, this means that the drop glue can in fact just not touch the panic count, as the panic count is already set by the in-progress unwind. It is *currently the case* that a panic-in-panic will abort [via the panic_count check](https://github.com/rust-lang/rust/blob/7b46aa594c4bdc507fbd904b6777ca30c37a9209/library/std/src/panicking.rs#L707-L714) in [`rust_panic_with_hook`](https://github.com/rust-lang/rust/blob/7b46aa594c4bdc507fbd904b6777ca30c37a9209/library/std/src/panicking.rs#L658).

It is the case that calling `drop_in_place` while panicking *cannot* unwind due to a rust panic. The only way that the separate landing pad would landed on is if an external unwind is caught. (IIUC, the intent is for `extern "C-unwind"` to convert allowed unwinds into a rust panic. This **must** involve incrementing the panic count (and thus a place to abort on double panic), as catching will decrement the panic count.)

If you can show a case, where the drop glue second landing pad is landed on, I'll happily be proven wrong. But the [normal case](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=779cc9aea8f0f2e5fc786f3ae6205161) is in fact that the abort comes *immediately* from the `panic!`, not the drop glue catching it.

<details><summary>So here's my updated drop glue impl:</summary>

```rust
// Codegen with -Z panic-in-drop=unwind (necessary behavior, different from current)
unsafe fn drop_in_place<T>(ptr: *mut T) {
    let unwind;

    // Note that try {} catch {} here is the raw intrinsics::try.
    // The panic count is not touched, so remains set from the first panic.

    // Call the Drop impl if there is one.
    try {
        <T as Drop>::drop(&mut *ptr);
    } catch(payload) {
        unwind = payload;
        goto 'unwind_field1;
    }

    // Drop the first field.
    try {
        drop_in_place(&mut (*ptr).field1);
    } catch(payload) {
        unwind = payload;
        goto 'unwind_field2;
    }

    // Drop the second field.
    try {
        drop_in_place(&mut (*ptr).field2);
    } catch(payload) {
        unwind = payload;
        goto 'unwind_field3;
    }

    // Drop the third field.
    try {
        drop_in_place(&mut (*ptr).field3);
    } catch(payload) {
        unwind = payload;
        goto 'done;
    }
    
    try {
    'unwind_field1:
        drop_in_place(&mut (*ptr).field1);
    'unwind_field2:
        drop_in_place(&mut (*ptr).field1);
    'unwind_field3:
        drop_in_place(&mut (*ptr).field1);
    } catch {
        // A rust panic cannot hit this landing pad, because it would've aborted.
        __rust_foreign_exception();
    }

'done:

    // Resume an in-progress unwind.
    if magic::initialized!(unwind) {
        __rust_resume_panic(unwind);
    }
}
```

</details>